phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -4168,4 +4176,197 @@ public MutationState useSchema(UseSchemaStatement useSchemaStatement) throws SQL
+
+        try {
+            if (grantStatement.getSchemaName() != null) {
+                // SYSTEM.CATALOG doesn't have any entry for "default" HBase namespace, hence we will bypass the check
+                if(!grantStatement.getSchemaName().equals(QueryConstants.HBASE_DEFAULT_SCHEMA_NAME)) {
+                    FromCompiler.getResolverForSchema(grantStatement.getSchemaName(), connection);
+                }
+                grantPermissionsToSchema(hConnection, grantStatement);
+
+            } else if (grantStatement.getTableName() != null) {
+                PTable inputTable = PhoenixRuntime.getTable(connection,
+                        SchemaUtil.normalizeFullTableName(grantStatement.getTableName().toString()));
+                if (!(PTableType.TABLE.equals(inputTable.getType()) || PTableType.SYSTEM.equals(inputTable.getType()))) {
+                    throw new AccessDeniedException("Cannot GRANT permissions on INDEX TABLES or VIEWS");
+                }
+                grantPermissionsToTables(hConnection, grantStatement, inputTable);
+
+            } else {
+                grantPermissionsToUser(hConnection, grantStatement);
+            }
+
+        } catch (SQLException e) {
+            // Bubble up the SQL Exception
+            throw e;
+        } catch (Throwable throwable) {
+            // Wrap around other exceptions to PhoenixIOException (Ex: org.apache.hadoop.hbase.security.AccessDeniedException)
+            throw ServerUtil.parseServerException(throwable);
+        }
+
+        return new MutationState(0, 0, connection);
+    }
+
+    private void grantPermissionsToTables(HConnection hConnection, GrantStatement grantStatement, PTable inputTable) throws Throwable {
+
+        org.apache.hadoop.hbase.TableName tableName = SchemaUtil.getPhysicalTableName
+                (inputTable.getName().getBytes(), inputTable.isNamespaceMapped());
+
+        grantPermissionsToTable(hConnection, grantStatement, tableName);
+
+        for(PTable indexTable : inputTable.getIndexes()) {
+            // Local Indexes don't correspond to new physical table, they are just stored in separate CF of base table.
+            if(indexTable.getIndexType().equals(IndexType.LOCAL)) {
+                continue;
+            }
+            logger.info("Granting " + Arrays.toString(grantStatement.getPermsList()) +
+                    " perms to IndexTable: " + indexTable.getName() + " BaseTable: " + inputTable.getName());
+            if (inputTable.isNamespaceMapped() != indexTable.isNamespaceMapped()) {
+                throw new TablesNotInSyncException(inputTable.getTableName().getString(),
+                        indexTable.getTableName().getString(), "Namespace properties");
+            }
+            tableName = SchemaUtil.getPhysicalTableName(indexTable.getName().getBytes(), indexTable.isNamespaceMapped());
+            grantPermissionsToTable(hConnection, grantStatement, tableName);
```

Comment:
you can have a common method to give you set of physical tables on which revoke and grant is required.

Reply:
Yeah that's a good idea. Let me refactor the code.